tests/test_helper.hpp

```diff
@@ -0,0 +1,193 @@
+}
+
+bool almost_equal(std::complex<float> x, sycl::ext::cplx::complex<float> y,
+                  int ulp) {
+  return almost_equal(x.real(), y.real(), ulp) &&
+         almost_equal(x.imag(), y.imag(), ulp);
+}
+
+bool almost_equal(std::complex<float> x, std::complex<float> y, int ulp) {
+  return almost_equal(x.real(), y.real(), ulp) &&
+         almost_equal(x.imag(), y.imag(), ulp);
+}
+
+bool almost_equal(sycl::ext::cplx::complex<sycl::half> x,
+                  sycl::ext::cplx::complex<sycl::half> y, int ulp) {
+  return almost_equal(x.real(), y.real(), ulp) &&
+         almost_equal(x.imag(), y.imag(), ulp);
+}
+
+bool almost_equal(sycl::ext::cplx::complex<sycl::half> x,
+                  std::complex<sycl::half> y, int ulp) {
+  return almost_equal(x.real(), y.real(), ulp) &&
+         almost_equal(x.imag(), y.imag(), ulp);
+}
+
+bool almost_equal(std::complex<sycl::half> x,
+                  sycl::ext::cplx::complex<sycl::half> y, int ulp) {
+  return almost_equal(x.real(), y.real(), ulp) &&
+         almost_equal(x.imag(), y.imag(), ulp);
+}
+
+bool almost_equal(std::complex<sycl::half> x, std::complex<sycl::half> y,
+                  int ulp) {
+  return almost_equal(x.real(), y.real(), ulp) &&
+         almost_equal(x.imag(), y.imag(), ulp);
+}
+
+// Helpers for testing half
+
+std::complex<float> sycl_half_to_float(sycl::ext::cplx::complex<sycl::half> c) {
+  auto c_sycl_float = static_cast<sycl::ext::cplx::complex<float>>(c);
+  return static_cast<std::complex<float>>(c_sycl_float);
+}
+
+std::complex<sycl::half> sycl_float_to_half(std::complex<float> c) {
+  auto c_sycl_half = static_cast<sycl::ext::cplx::complex<sycl::half>>(c);
+  return static_cast<std::complex<sycl::half>>(c_sycl_half);
+}
+
+std::complex<float> trunc_float(std::complex<float> c) {
+  auto c_sycl_half = static_cast<sycl::ext::cplx::complex<sycl::half>>(c);
+  return sycl_half_to_float(c_sycl_half);
+}
+
+template <typename T_in> auto constexpr init_std(T_in re, T_in im) {
+  return std::complex<T_in>(re, im);
+}
+
+template <> auto constexpr init_std(sycl::half re, sycl::half im) {
+  return trunc_float(std::complex<float>(re, im));
+}
+
+template <typename T_in> auto constexpr init_deci(T_in re) { return re; }
+
+template <> auto constexpr init_deci(sycl::half re) {
+  return static_cast<float>(re);
+}
+
+// Helpers for comparing SyclCPLX and standard c++ results
+
+template <typename T>
+bool check_results(sycl::ext::cplx::complex<T> output,
+                   std::complex<T> reference, bool is_device) {
+  if (!almost_equal(output, reference, SYCL_CPLX_TOL_ULP)) {
+    std::cerr << std::setprecision(std::numeric_limits<T>::max_digits10)
+              << "Test failed with complex_type: " << get_typename<T>()
+              << " Computed on " << (is_device ? "device" : "host")
+              << " Output: (" << output.real() << "," << output.imag() << ")"
+              << " Reference: (" << reference.real() << "," << reference.imag()
```

Comment:
Can we rely on the `<<` overload of `std::complex` and `sycl::ext::cplx::complex`  so we don't need to use `.real()` / `imag()` here?

Reply:
Yes, I have changed this to use the `<<` overload.